### ISSUE TITLE
Parallelization control of DynamicNode

### ISSUE THREAD

**Issue by big-andy-coates** [junit-team/junit-framework#2497]
`@TestFactory` can return a stream of `DynamicContainers` and `DynamicTest`.  It would be awesome if we could run containers in parallel, but the tests within containers on the same thread.

The `@Execution` seems to only allow you to switch from everything parallel/concurrent or everything on the same thread.

The configuration parameters `junit.jupiter.execution.parallel.mode.default` and  `junit.jupiter.execution.parallel.mode.classes.default` also don't seem to allow such a configuration, probably because all test suites and test are coming from the same `class`.

For example: the following should allow tests `test1` and `test2` that are in `suite1` to be run on the same thread, and the tests in `suite2` to be run on a different, but consistent thread:

```java
static class Runner {

    @TestFactory
    Stream<DynamicNode> systemTest() {
        final Stream<DynamicNode> tests1 = Stream.of(
                        DynamicTest.dynamicTest("test1", () -> {System.out.println(Thread.currentThread().getId() + " -test1");}),
                        DynamicTest.dynamicTest("test2", () -> {System.out.println(Thread.currentThread().getId() + " - test2");}));

        final Stream<DynamicNode> tests2 = Stream.of(
                        DynamicTest.dynamicTest("test3", () -> {System.out.println(Thread.currentThread().getId() + " -test3");}),
                        DynamicTest.dynamicTest("test4", () -> {System.out.println(Thread.currentThread().getId() + " - test4");}));

        return Stream.of(
                DynamicContainer.dynamicContainer("suite1", tests1),
                DynamicContainer.dynamicContainer("suite2", tests2));
    }
}
```

Maybe this could be achieved via one or more of:
  * a new `junit.jupiter.execution.parallel.mode.container.default` property
  * a field on `@TestFactory` to control execution
  * an enhancement to `@Execution`
  * a parameter to `DynamicContainer`

Thanks!

**Comment by marcphilipp** [comment]
Tentatively slated for 5.8 M1 solely for the purpose of _team discussion_.

**Comment by sic-poths** [comment]
We have a similar case, but would like to run the first level of dynamic nodes in sequence, but tests under each level can be run in parallel. It would be great to be able to define the `ExecutionMode` on each `DynamicContainer` instance.

**Comment by Mincol** [comment]
I would like to see 4 executables emulating @BeforeAll, @BeforeEach , @AfterEach and @AfterAll

**Comment by sic-poths** [comment]
I would still be really interested in a solution to this problem. Are there any thoughts from the team on what a good implementation of a more fine granular "execution-order-control" for dynamic tests could be? Maybe I would give it a try to create a PR, but I only know the codebase as a user.

**Comment by niteshhardikar** [comment]
Perfectly explained by @big-andy-coates. I have run into the same requirement where I want to run my DynamicContainers in parallel but the DynamicTests within each container sequentially. Is this slated for any upcoming release or is there any other solution to this problem for instance using the Extension model? Any inputs are highly appreciated.

**Comment by marcphilipp** [comment]
I wonder if a configuration parameter would really be adequate here. Wouldn't it be better if `DynamicNode` would provide an API to configure the execution mode similiar to `@Execution` for regular test classes and methods?

@big-andy-coates @sic-poths @niteshhardikar WDYT?

**Comment by big-andy-coates** [comment]
Yep, that's certainly one option. I mentioned in the original text about adding "a parameter to DynamicContainer", though as you note, `DynamicNode` would be better.

**Comment by niteshhardikar** [comment]
Yes @marcphilipp. Supporting configuration of execution mode for DynamicNode would be ideal.

**Comment by marcphilipp** [comment]
@big-andy-coates @niteshhardikar Do you have concrete proposals for that API? I'm reluctant to add more overloads for `DynamicTest.dynamicTest` and `DynamicContainer.dynamicContainer`. 🤔

**Comment by big-andy-coates** [comment]
Maybe it's time to introduce a builder pattern, rather than overloading?

```java
DynamicContainer.builder("displayName")
   .withNodes(iterable)
   .withNodes(stream)
   .withTestSource(uri)
   .withExecutionMode(mode)
   .build();
```

**Comment by marcphilipp** [comment]
The methods in this codebase don't use the `with` prefix (see `LauncherDiscoveryRequestBuilder`). Apart from that I can see an API like that work.

This recent PR is also related: https://github.com/junit-team/junit5/pull/3220#pullrequestreview-1401371933

**Comment by marcphilipp** [comment]
**Team decision:** While we think this is a reasonable request, we'd like to wait for additional interest from the community

**Comment by novoj** [comment]
What can the interested community do to move this forward? Are you waiting for a certain number of votes/responses here or how "the interest" is measured?

**Comment by marcphilipp** [comment]
Yes, interest is measured in the number of upvotes.

**Comment by prathasirisha** [comment]
I have upvoted 👍🏼 for the issue and the use case outlined here: https://github.com/junit-team/junit5/issues/2497#issuecomment-766923020

**Comment by sonthanh** [comment]
+1

**Comment by johnedquinn** [comment]
I see that this issue is tied for #3 on your list of sorted issues by number of upvotes. Has there been any traction on this?

**Comment by marcphilipp** [comment]
The above API proposal (see https://github.com/junit-team/junit5/issues/2497#issuecomment-1519097081) didn't sit well with me, so I spent some time this morning to try a different approach. It looks as follows (see draft PR #4621 for details):

```java
dynamicContainer(root -> root
	.displayName("container")
	.source(URI.create("other://container"))
	.executionMode(ExecutionMode.CONCURRENT)
	.defaultChildExecutionMode(ExecutionMode.SAME_THREAD, "optional reason")
	.children(
		dynamicTest(child -> child
			.displayName("test1")
			.source(URI.create("any://test1"))
			.executionMode(ExecutionMode.CONCURRENT)
			.executable(() -> {
				// do something
				assertEquals(2, 1 + 1);
			})
		),
		dynamicTest(child -> child
			.displayName("test2")
			.source(URI.create("any://test2"))
			.executable(() -> {
				// do something
				assertEquals(2, 1 + 1);
			})
		)
	)
);
```

WDYT?

**Comment by johnedquinn** [comment]
Thank you @marcphilipp! This would totally solve my issue. This looks like the builder pattern, but not entirely. I'm personally inclined to use the builder pattern, but I'm really just looking for the functionality that this represents. Hopefully the JUnit team can work it out.

**Comment by Janrupf** [comment]
I'd also be interested in this. I would love this to also support streams, though my case would also work without streams.

**Comment by marcphilipp** [comment]
> I'd also be interested in this. I would love this to also support streams, though my case would also work without streams.

The draft PR has this method:

```java
Configuration children(Stream<? extends DynamicNode> children);
```

It would work like this:

```java
dynamicContainer(root -> root
	.displayName("container")
	.children(Stream.of(dynamicTest(...), dynamicTest(...))));
```

@Janrupf Is that what you meant?

**Comment by Janrupf** [comment]
I think that would do, yes
